src/Ens.ts

```diff
@@ -0,0 +1,594 @@
+  readonly proxyServiceApiKey: string | undefined;
+
+  constructor(source?: EnsSource) {
+    super();
+    let finalSource: EnsSource = {url: '', network: 'mainnet'};
+    if (source) {
+      finalSource = this.checkNetworkConfig(source);
+    }
+    this.network = EthereumNetworks[finalSource.network];
+    this.networkName = finalSource.network;
+    this.url = finalSource['url'];
+    this.provider =
+      finalSource['provider'] ||
+      FetchProvider.factory(NamingServiceName.ENS, this.url);
+    this.proxyServiceApiKey = finalSource['proxyServiceApiKey'];
+
+    const registryAddress =
+      finalSource['registryAddress'] || EnsNetworkMap[this.network];
+    this.registryContract = new EthereumContract(
+      ensInterface,
+      registryAddress,
+      this.provider,
+      this.proxyServiceApiKey,
+    );
+
+    const nameWrapperAddress = this.determineNameWrapperAddress(this.network);
+    this.nameWrapperContract = new EthereumContract(
+      nameWrapperInterface,
+      nameWrapperAddress,
+      this.provider,
+      this.proxyServiceApiKey,
+    );
+
+    const baseRegistrarAddress = this.determineBaseRegistrarAddress(
+      this.network,
+    );
+    this.baseRegistrarContract = new EthereumContract(
+      baseRegistrarInterface,
+      baseRegistrarAddress,
+      this.provider,
+      this.proxyServiceApiKey,
+    );
+  }
+
+  static async autoNetwork(
+    config: {url: string} | {provider: Provider},
+  ): Promise<Ens> {
+    let provider: Provider;
+
+    if (hasProvider(config)) {
+      provider = config.provider;
+    } else {
+      if (!config.url) {
+        throw new ConfigurationError(ConfigurationErrorCode.UnspecifiedUrl, {
+          method: NamingServiceName.ENS,
+        });
+      }
+      provider = FetchProvider.factory(NamingServiceName.ENS, config.url);
+    }
+
+    const networkId = (await provider.request({
+      method: 'net_version',
+    })) as number;
+    const networkName = EthereumNetworksInverted[networkId];
+    if (!networkName || !EnsSupportedNetwork.guard(networkName)) {
+      throw new ConfigurationError(ConfigurationErrorCode.UnsupportedNetwork, {
+        method: NamingServiceName.ENS,
+      });
+    }
+    return new this({network: networkName, provider: provider});
+  }
+
+  serviceName(): NamingServiceName {
+    return this.name;
+  }
+
+  namehash(domain: string): string {
+    if (!this.checkSupportedDomain(domain)) {
+      throw new ResolutionError(ResolutionErrorCode.UnsupportedDomain, {
+        domain,
+      });
+    }
+    return eip137Namehash(domain);
+  }
+
+  childhash(parentHash: string, label: string): string {
+    return eip137Childhash(parentHash, label);
+  }
+
+  async isSupportedDomain(domain: string): Promise<boolean> {
+    return this.checkSupportedDomain(domain);
+  }
+
+  async owner(domain: string): Promise<string> {
+    const namehash = this.namehash(domain);
+    return await this.callMethod(this.registryContract, 'owner', [namehash]);
```

Comment:
this missing the onwner address of namewrapper address
See https://github.com/unstoppabledomains/resolution-go/blob/master/ens_service.go#L156-L186
you can also check with namewrapper contract isWrapped API first

Reply:
ah looks like you already handle this somewhere? I saw your test result from linear

Reply:
Im getting confused what you're saying are you talking about the namewrapper address.

`  /**
   * @param domain - domain name
   * @returns An owner address of the domain
   */
  async owner(domain: string): Promise<string | null> {
    domain = prepareAndValidateDomain(domain);
    return this.callServiceForDomain(domain, (service) =>
      service.owner(domain),
    );
  }`